### PR TITLE
Add OpenApiRefProcessor for handling $ref in OpenAPI specs

### PR DESCRIPTION
**Pull request by mfrinnstrom**
Signed-off-by: Mattias Frinnström <mattias@frinnstrom.se>

## Hey, I just made a Pull Request!

Added basic support for handling `$ref` in OpenAPI specifications. I added it as a separate, optional, processor that bundles the specification into one during processing. I'm guessing there are cases that are not covered but my [example repo](https://github.com/mfrinnstrom/backstage-openapi-ref-example) is working.

The bundling works by replacing the normal file resolver of `swagger-parser` with one that uses an `UrlReader` for fetching files.

This will hopefully be a start for solving #6832.

#### :heavy_check_mark: Checklist

- [X] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [X] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [X] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

**Comment by mfrinnstrom** [comment]
I'm not sure what to do about the `openapi-types` error. I tried adding it to `devDependencies` but then I got an error with `minipass` instead. 

There seems to be an open issue in `swagger-parser` about this so maybe I have to wait for that. https://github.com/APIDevTools/swagger-parser/issues/199

**Comment by benjdlambert** [comment]
What's the error with minipass that you get?

**Comment by mfrinnstrom** [comment]
When I add `openapi-types` as a dev dependency and run `yarn tsc:full` I get the following errors:

```
node_modules/@types/minipass/index.d.ts:27:5 - error TS2416: Property 'end' in type 'MiniPass' is not assignable to the same property in base type 'WritableStream'.
  Type '{ (cb?: (() => void) | undefined): void; (chunk: any, cb?: (() => void) | undefined): void; (chunk: any, encoding?: string | null | undefined, cb?: (() => void) | undefined): void; }' is not assignable to type '{ (cb?: (() => void) | undefined): this; (data: string | Uint8Array, cb?: (() => void) | undefined): this; (str: string, encoding?: BufferEncoding | undefined, cb?: (() => void) | undefined): this; }'.
    Type 'void' is not assignable to type 'this'.
      'this' could be instantiated with an arbitrary type which could be unrelated to 'void'.

27     end(cb?: () => void): void;
       ~~~

node_modules/@types/minipass/index.d.ts:28:5 - error TS2416: Property 'end' in type 'MiniPass' is not assignable to the same property in base type 'WritableStream'.
  Type '{ (cb?: (() => void) | undefined): void; (chunk: any, cb?: (() => void) | undefined): void; (chunk: any, encoding?: string | null | undefined, cb?: (() => void) | undefined): void; }' is not assignable to type '{ (cb?: (() => void) | undefined): this; (data: string | Uint8Array, cb?: (() => void) | undefined): this; (str: string, encoding?: BufferEncoding | undefined, cb?: (() => void) | undefined): this; }'.

28     end(chunk: any, cb?: () => void): void;
       ~~~

node_modules/@types/minipass/index.d.ts:29:5 - error TS2416: Property 'end' in type 'MiniPass' is not assignable to the same property in base type 'WritableStream'.
  Type '{ (cb?: (() => void) | undefined): void; (chunk: any, cb?: (() => void) | undefined): void; (chunk: any, encoding?: string | null | undefined, cb?: (() => void) | undefined): void; }' is not assignable to type '{ (cb?: (() => void) | undefined): this; (data: string | Uint8Array, cb?: (() => void) | undefined): this; (str: string, encoding?: BufferEncoding | undefined, cb?: (() => void) | undefined): this; }'.

29     end(chunk: any, encoding?: string | null, cb?: () => void): void;
       ~~~


Found 3 errors in the same file, starting at: node_modules/@types/minipass/index.d.ts:27
```

It might have been there before just that it exited on the missing `openapi-types` then.

**Review by jhaals — COMMENTED** [review]
Thanks for looking at this Mattias!

Given that this is a new processor and that we already started moving entity providers and processors out into separate modules I think we should do the same for this one.
We don't have tooling for creating catalog modules so you can either copy an existing catalog module from `plugins/` or run `yarn backstage-cli create` select scaffolder module and use that as your template.

**Comment by mfrinnstrom** [comment]
@jhaals Thanks for the input. 

Do you have any thoughts around the name for the module? Should it be `catalog-backend-module-openapi` or something more generic like `catalog-backend-module-api`?

**Review by jhaals — COMMENTED** [review]
Your `catalog-backend-module-openapi` suggestion sounds good 😄 

The actual name in package.json will be `@backstage/plugin-catalog-backend-module-openapi` but the folder name in the plugins folder will be `catalog-backend-module-openapi`. There are already some existing modules for inspiration

**Comment by mfrinnstrom** [comment]
@jhaals I have moved it to a separate module now. Hopefully I managed to get everything right.

File: plugins/catalog-backend-module-openapi/package.json [hunk 1]
```diff
@@ -0,0 +1,53 @@
+{
+  "name": "@backstage/plugin-catalog-backend-module-openapi",
+  "description": "A Backstage catalog backend module that helps with OpenAPI specifications",
+  "version": "0.1.0",
```
Comment (jhaals) [inline plugins/catalog-backend-module-openapi/package.json:4]: Make this `0.0.0`, the changeset will make sure the first version released will be 0.1.0

**Review by jhaals — COMMENTED** [review]
One small thing to fix otherwise 👍 in my book!

File: plugins/catalog-backend-module-openapi/package.json (hunk 1 above)
Comment (mfrinnstrom) [reply plugins/catalog-backend-module-openapi/package.json]: Updated now.

**Review by jhaals — APPROVED** [review]
Nice work! 👍

**Pull request merged**